README.md
# Virtuemart 3 - Xendit Plugin

## How to Install

## Prerequisite
- Install [Joomla](https://devilbox.readthedocs.io/en/latest/examples/setup-joomla.html) via Devilbox
- Install [VirtueMart](https://virtuemart.net/downloads) extension or download from Joomla marketplace

### Manually
To install this plugin manually, there are two main steps that you need to do:

#### Download and upload
1. Download [the latest release](https://github.com/xendit/virtuemart-module/releases/tag/1.2.0).
2. On your store's administrator page, go to `Extensions -> Install` page
3. Select `Upload Package File`
4. Upload the downloaded file from step 1
5. You should get a message saying that your installation is successful

#### Activate
1. You should see `VM Payment - Xendit` in the list of extension
2. Select `VM Payment - Xendit` and click `Install` button
3. Go to `Extensions -> Manage` page
4. Type `xendit` in the search bar
5. Select `VM Payment - Xendit` and click `Enable` button

## How to Use
### Create new Xendit payment method
1. On your store's administrator page, go to `Components -> VirtueMart -> Payment Method` page
2. Click the `New` button
3. Fill the form:
- Payment Name -> This will be shown to your customer in the checkout page. E.g. `Xendit`
- Description -> This will be shown to your customer in the checkout page. E.g. `Pay via Xendit`
- Payment Method -> `VM Payment - Xendit`
- Currency -> `Indonesian Rupiah`. Currently we only support IDR as the currency

## Ownership

Team: [TPI Team](https://www.draw.io/?state=%7B%22ids%22:%5B%221Vk1zqYgX2YqjJYieQ6qDPh0PhB2yAd0j%22%5D,%22action%22:%22open%22,%22userId%22:%22104938211257040552218%22%7D)

Slack Channel: [#p-integration](https://xendit.slack.com/messages/p-integration)

Slack Mentions: `@troops-tpi`


CHANGELOG.md
# CHANGELOG
## 1.2.0 (2021-09-24)
- Change cc implementation using Invoice flow

## 1.2.1 (2022-02-16)
- Improve item

## 1.1.0 (2021-01-19)
- Error message standardization

## 1.0.2 (2020-06-03)
- Remove unused month and year

## 1.0.1 (2020-05-08)
- Reconfigure Dynamic 3DS

## 1.0.0 (2020-03-03)
- Add support to choose different payment method
- Support these payment methods:
  1. CC
  2. BCA Virtual Accounts (VA)
  3. BNI VA
  4. BRI VA
  5. Mandiri VA
  6. Permata VA
- Implement notification endpoint for changing order status to "Confirmed by Shopper" once the payment is successful

## 0.0.1 (2020-02-27)

Features:
- Available payment method: BNI
- Minimum and maximum amount validation


xendit_callback.php
<?php
/**
 * This is file for handle callback
 * @link baseURL() /plugins/vmpayment/xendit/xendit_callback.php?xendit_mode=[xendit_mode]
 * xendit_mode:
 *      - xendit_invoice_callback
 *      - xendit_ewallet_callback (not support yet)
 *      - xendit_cardless_callback (not support yet)
 *      - xendit_cc_callback (not support yet)
 */

$ch = curl_init();

$end_point = 'https://tpi.xendit.co/payment/xendit/invoice';

$data = array(
    'external_id' => 'demo_1475801962607',
    'amount' => 230000,
    'payer_email' => 'candra@xendit.co',
    'description' => 'Jalan jalan'
);

$post_string = json_encode($data);

curl_setopt($ch, CURLOPT_URL, $end_point);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
curl_setopt($ch, CURLOPT_POSTFIELDS, $post_string);
curl_setopt($ch, CURLOPT_HEADER, 0);
curl_setopt($ch, CURLOPT_TIMEOUT, 45);
curl_setopt($ch, CURLOPT_POST, 1);

curl_setopt($ch, CURLOPT_HTTPHEADER, array(
    'Authorization: Basic ' . base64_encode('x:'),
    'content-type: application/json',
    'x-plugin-name: VIRTUEMART'
));

$response = curl_exec($ch);

// var_dump(json_decode(json_encode(curl_getinfo($ch))));

curl_close($ch);

var_dump($response);


xendit/tmpl/cc_payment.php
<?php
defined ('_JEXEC') or die();

vmJsApi::css('xendit','plugins/vmpayment/xendit/xendit/assets/css/');
vmJsApi::addJScript('https://cdnjs.cloudflare.com/ajax/libs/jquery.payment/3.0.0/jquery.payment.min.js');
vmJsApi::addJScript('https://js.xendit.co/v1/xendit.min.js');
?>

<input type="radio" name="virtuemart_paymentmethod_id" class="xendit-pg-input"
       id="payment_id_<?php echo $viewData['plugin']->virtuemart_paymentmethod_id; ?>"
	   value="<?php echo $viewData['plugin']->virtuemart_paymentmethod_id; ?>" <?php echo $viewData ['checked']; ?>
	   xendit-payment-type="<?php echo $viewData['plugin']->xendit_gateway_payment_type; ?>">
<label for="payment_id_<?php echo $viewData['plugin']->virtuemart_paymentmethod_id; ?>">
    <span class="vmpayment" style="vertical-align: middle;">
        <?php if (!empty($viewData['payment_logo'])) { ?>
	        <span class="vmCartPaymentLogo"><?php echo $viewData ['payment_logo']; ?></span>
        <?php } ?>
	    <span class="vmpayment_name"><?php echo $viewData['plugin']->payment_name; ?></span>
	    <?php if (!empty($viewData['plugin']->payment_desc)) { ?>
		    <span class="vmpayment_description"><?php echo $viewData['plugin']->payment_desc; ?></span>
	    <?php } ?>
	    <?php if (!empty($viewData['payment_cost'])) { ?>
		    <span class="vmpayment_cost"><?php echo vmText::_ ('COM_VIRTUEMART_PLUGIN_COST_DISPLAY') . $viewData['payment_cost']; ?></span>
	    <?php } ?>
    </span>
</label>


xendit/tmpl/post_payment.php
<?php
defined ('_JEXEC') or die();
?>
<div class="post_payment_order_number" style="width: 100%">
	<span class="post_payment_order_number_title"><?php echo vmText::_ ('COM_VIRTUEMART_ORDER_NUMBER'); ?> </span>
	<?php echo $viewData['order']['details']['BT']->order_number; ?>
</div>
<?php if ($viewData['paymentInfos']->sofort_response_transaction)  { ?>
	<div class="post_payment_transaction" style="width: 100%">
	<span class="post_payment_transaction_title"><?php echo vmText::_ ('VMPAYMENT_SOFORT_RESPONSE_TRANSACTION'); ?> </span>
<?php echo $viewData['paymentInfos']->sofort_response_transaction; ?>
</div>
<?php } ?>

<div class="post_payment_order_total" style="width: 100%">
	<span class="post_payment_order_total_title"><?php echo vmText::_ ('COM_VIRTUEMART_ORDER_PRINT_TOTAL'); ?> </span>
	<?php echo $viewData['displayTotalInPaymentCurrency']; ?>
</div>
<a class="vm-button-correct" href="<?php echo JRoute::_('index.php?option=com_virtuemart&view=orders&layout=details&order_number='.$viewData["order"]['details']['BT']->order_number.'&order_pass='.$viewData["order"]['details']['BT']->order_pass, false)?>"><?php echo vmText::_('COM_VIRTUEMART_ORDER_VIEW_ORDER'); ?></a>









xendit/tmpl/display_payment.php
<?php
defined ('_JEXEC') or die();

vmJsApi::css('xendit','plugins/vmpayment/xendit/xendit/assets/css/');
?>

<input type="radio" name="virtuemart_paymentmethod_id" class="xendit-pg-input"
       id="payment_id_<?php echo $viewData['plugin']->virtuemart_paymentmethod_id; ?>"
	   value="<?php echo $viewData['plugin']->virtuemart_paymentmethod_id; ?>" <?php echo $viewData ['checked']; ?>
	   xendit-payment-type="<?php echo $viewData['plugin']->xendit_gateway_payment_type; ?>">
<label for="payment_id_<?php echo $viewData['plugin']->virtuemart_paymentmethod_id; ?>">
    <span class="vmpayment" style="vertical-align: middle;">
        <?php if (!empty($viewData['payment_logo'])) { ?>
	        <span class="vmCartPaymentLogo"><?php echo $viewData ['payment_logo']; ?></span>
        <?php } ?>
	    <span class="vmpayment_name"><?php echo $viewData['plugin']->payment_name; ?></span>
	    <?php if (!empty($viewData['plugin']->payment_desc)) { ?>
		    <span class="vmpayment_description"><?php echo $viewData['plugin']->payment_desc; ?></span>
	    <?php } ?>
	    <?php if (!empty($viewData['payment_cost'])) { ?>
		    <span class="vmpayment_cost"><?php echo vmText::_ ('COM_VIRTUEMART_PLUGIN_COST_DISPLAY') . $viewData['payment_cost']; ?></span>
	    <?php } ?>
    </span>
</label>

xendit/lib/xendit_api.php
<?php

defined('_JEXEC') or die('Restricted access');

/**
 * @mainpage
 * Base class for Xendit REST API
 * This class implements basic http authentication and JSON parse to PHP array
 * for the response
 *
 * Requires libcurl and openssl
 *
 * Copyright (c) 2020 Xendit
 *
 * Released under the GNU General Public License (Version 3)
 * [https://www.gnu.org/licenses/gpl-3.0.html]
 *
 * $Date: 2020-03-03 14:08:47 +0700 (Tue, 3 Mar 2020) $
 * @author Xendit https://xendit.co (thirdpartyintegrations@xendit.co)
 *
 */

class XenditApi {
    const METHOD_POST = 'POST';
    const METHOD_GET = 'GET';

    function __construct ($method) {
        $this->server_domain = 'https://api.xendit.co';
        $this->tpi_server_domain = 'https://tpi.xendit.co';

        $this->environment = $method->shop_mode ? $method->shop_mode : 'test';
        $this->payment_type = $method->xendit_gateway_payment_type;

        if ($this->payment_type) { // identified as xendit payment
            if (($this->environment=='test' && (empty($method->xendit_gateway_public_api_key_test) || empty($method->xendit_gateway_secret_api_key_test)))
            ||
            ($this->environment!='test' && (empty($method->xendit_gateway_public_api_key) || empty($method->xendit_gateway_secret_api_key)))){
                vmError(vmText::sprintf('VMPAYMENT_XENDIT_PARAMETER_REQUIRED'));

                return FALSE;
            }
        }

        $this->secret_api_key = $this->environment=='test' ? $method->xendit_gateway_secret_api_key_test : $method->xendit_gateway_secret_api_key;
        $this->public_api_key = $this->environment=='test' ? $method->xendit_gateway_public_api_key_test : $method->xendit_gateway_public_api_key;
    }

    function getHeader() {
        $site_config = JFactory::getConfig();
        $store_name = $site_config->get('sitename');

        return array(
            'content-type: application/json',
            'x-plugin-name: VIRTUEMART',
            'x-plugin-version: 1.2.1',
            'x-plugin-store-name: ' . $store_name
        );
    }

    function getPublicKey() {
        return $this->public_api_key;
    }

    /*******************************************************************************
        Virtual Accounts
     *******************************************************************************/
    function createInvoice ($body, $header) {
        $endpoint = $this->tpi_server_domain.'/payment/xendit/invoice';
        $default_header = $this->getHeader();
        $header = array_merge($header, $default_header);

        // return $header;

        $json_response = $this->_sendRequest($endpoint, self::METHOD_POST, $body, $header);

        return $json_response;
    }

    function getInvoice($invoice_id='') {
        $endPoint = $this->tpi_server_domain.'/payment/xendit/invoice/'.$invoice_id;
        $default_header = $this->getHeader();
        $body = [];

        $json_response = $this->_sendRequest($endPoint, self::METHOD_GET, $body, $default_header);

        return $json_response;
    }

    /*******************************************************************************
        Credit Card
     *******************************************************************************/

    function createCharge($body) {
        $endpoint = $this->tpi_server_domain.'/payment/xendit/credit-card/charges';
        $default_header = $this->getHeader();

        $json_response = $this->_sendRequest($endpoint, self::METHOD_POST, $body, $default_header);

        return $json_response;
    }
    
     function createHosted3DS($body, $header) {
        $endpoint = $this->tpi_server_domain.'/payment/xendit/credit-card/hosted-3ds';
        $default_header = $this->getHeader();
        $header = array_merge($header, $default_header);
        $options = array(
            'should_use_public_key' => true
        );

        $json_response = $this->_sendRequest($endpoint, self::METHOD_POST, $body, $header, $options);

        return $json_response;
    }

    /**
     * Get CC Setting
     * Note: the return will be array, but if value is boolean (true) json_decode will convert to "1" otherwise if value is boolean (false) json_decode will convert to ""
     */
    function getCCSettings() {
        $endpoint = $this->tpi_server_domain.'/payment/xendit/settings/credit-card';
        $default_header = $this->getHeader();
        $body = [];

        $json_response = $this->_sendRequest($endpoint, self::METHOD_GET, $body, $default_header);

        return $json_response;
    }

    function getCharge($charge_id='') {
        $endPoint = $this->tpi_server_domain.'/payment/xendit/credit-card/charges/'.$charge_id;
        $default_header = $this->getHeader();
        $body = [];

        $json_response = $this->_sendRequest($endPoint, self::METHOD_GET, $body, $default_header);

        return $json_response;
    }

    /**
	 * _sendRequest
	 * Posts the request to AuthorizeNet & returns response using curl
	 *
	 * @author Valerie Isaksen
	 * @param string $url
	 * @param string $content
	 *
	 */
	function _sendRequest($endpoint, $method, $body = array(), $header = array(), $options = array()) {
        $ch = curl_init();
        
        $should_use_public_key = isset($options['should_use_public_key']) ? $options['should_use_public_key'] : false;
        $api_key = $should_use_public_key ? $this->public_api_key : $this->secret_api_key;

        $curl_options = array(
            CURLOPT_URL => $endpoint,
            CURLOPT_HTTPHEADER => $header,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_USERPWD => $api_key . ':'
        );

        if ($method === self::METHOD_POST) {
            $curl_options[CURLOPT_POST] = true;
            $curl_options[CURLOPT_POSTFIELDS] = json_encode($body);
        }

        curl_setopt_array($ch, $curl_options);
        
        
        $response = curl_exec($ch);
        
        curl_close($ch);

        if ($response === false) {
            throw new Exception('Xendit cURL Error, error code: ' . curl_error($ch), curl_errno($ch));
        }

        $json_response = json_decode($response, true);

        return $json_response;
	}
}

?>